### PR TITLE
Update TTS calculation method

### PR DESCRIPTION
**Pull request by ChanWooY-1QBit**
Please see the [contributing guidelines](https://github.com/ccvm/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description

> This PR is to update the TTS calculation method. The new method does not use arbitrary values of `best_energy` and is faster.
> `def generate_plot_data()` is now renamed with `def generated_TTS_plot_data()`
> `TTS_type` is not an attribute of the `ProblemMetadata` anymore, and now it is one of the inputs of `def generate_TTS_plot_data()`. 
> Related testcases are updated

### References

> #128 

### Testing

> Describe how this can be tested by reviewers. Be specific about anything not tested and reasons why.
>
> Please include details of the environment this PR was developed in.

- [√ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [√ ] I have added documentation for new/changed functionality in this PR
- [ √] All active GitHub checks for tests, formatting, and security are passing
- [ √] The correct base branch is being used, if not `main`

### CODE REVIEW

File: ccvm_simulators/ccvmplotlib/problem_metadata/boxqp_metadata.py [hunk 1]
```diff
@@ -83,69 +81,26 @@ def ingest_metadata(self, metadata_filepath: str) -> None:
...
-        self, matching_df: pd.DataFrame, percent_gap: str, problem_size: int
-    ) -> list:
-        """A method to generate more data points that is used for plotting.
-
-        With a given result data in pandas.DataFrame, it generates best known
-        energy list, Time-To-Solve list, and success probability list based on
-        the optimal type and problem size inputs.
+    def generate_TTS_plot_data(
+        self,
+        method="advanced",
```
Comment (FarhadK-1QBit) [inline ccvm_simulators/ccvmplotlib/problem_metadata/boxqp_metadata.py:86]: The variable `method` is not used and it's not really needed. In all the calculations of TTS we should just use the "advanced" method. We can remove this variable.

File: ccvm_simulators/ccvmplotlib/problem_metadata/boxqp_metadata.py [hunk 2]
```diff
@@ -159,32 +114,73 @@ def generate_plot_data(self) -> pd.DataFrame:
...
+                        iterations = matching_df["iterations"].values
+                        roundtrip_time = (
+                            (
+                                device_parameters["FPGA_fixed"]
+                                + device_parameters["FPGA_var_fac"]
+                                * float(problem_size)
+                            )
+                            * device_parameters["FPGA_clock"]
+                            + float(problem_size) * device_parameters["laser_clock"]
+                            + device_parameters["buffer_time"]
```
Comment (FarhadK-1QBit) [inline ccvm_simulators/ccvmplotlib/problem_metadata/boxqp_metadata.py:136]: This calculation of `roundtrip_time` is specific to the MF-CCVM solver. For the DL-CCVM solver, it should be written as:
```
roundtrip_time = float(problem_size) * device_parameters["laser_clock"]
```

**Review by FarhadK-1QBit — COMMENTED** [review]
Thank you for these changes, Chan Woo! 

I'm wondering if we need to make some changes to the calculation of TTS (which would also affect the ETS calculation).
As we discussed, maybe we can remove the variable `TTS_type` and replace it with a variable named `machine` or `device` which, when it's equal to `dl-ccvm` or `mf-ccvm` it would calculate the TTS according to the `roundtrip_time` calculation I mentioned in my comment in this review. 
I will post a detailed review of this in the ccvm-boxqp channel.

File: ccvm_simulators/ccvmplotlib/problem_metadata/boxqp_metadata.py (hunk 2 above)
Comment (ChanWooY-1QBit) [reply ccvm_simulators/ccvmplotlib/problem_metadata/boxqp_metadata.py:133]: Does this mean that when we generate TTS data for plotting, the plotter needs to identify which solver was used because the `rounttrip_time` calculation depends on it?

Comment (ChanWooY-1QBit) [reply ccvm_simulators/ccvmplotlib/problem_metadata/boxqp_metadata.py]: This `machine_time` calculation is replaced with a callback function. Thus, the `roundtrip_time` will be moved into the callback function.

File: examples/ccvm_boxqp_plot.py [hunk 3]
```diff
@@ -50,10 +50,17 @@
...
+    # Customize machine_parameters
+    machine_parameters = {
+        "cpu_power": {20: 5.0, 30: 5.0, 40: 5.0, 50: 5.0, 60: 5.0, 70: 5.0}
+    }
+
     plot_fig, plot_ax = ccvmplotlib.plot_TTS(
         metadata_filepath=metadata_filepath,
         problem="BoxQP",
-        TTS_type="wallclock",
+        machine_time_func=solver.cpu_machine_time(
```
Comment (KatieO4t1qbit) [inline examples/ccvm_boxqp_plot.py:61]: I believe this won't run until we have the solver bits written, too, and we don't want to merge something that can't be run. However, I like that this PR is staying within scope, thank you.
My recommendation to merge this fast is to add an example of how the user can define their own cpu_machine_time() function _in this file_. As long as this file runs, the exact implementation of that function doesn't matter, but for the sake of making the best code we can let's use the cpu implementation.

**Review by KatieO4t1qbit — CHANGES_REQUESTED** [review]

**Review by FarhadK-1QBit — APPROVED** [review]
These look good to me, Chan Woo! Thank you!

File: ccvm_simulators/ccvmplotlib/tests/test_BoxQP_metadata.py [hunk 4]
```diff
@@ -35,52 +49,69 @@ def test_BoxQP_metadata_invalid_problem_type(self):
...
-
-        """
-        with self.assertRaises(ValueError):
-            BoxQPMetadata(self.valid_problem_type, TTSType(self.invalid_TTS_type_str))
-
     def test_ingest_result_data_method_valid(self):
         """Test BoxQP Metadata class "ingest_result_data" method when valid
         inputs are given.
 
         # TODO: Implementation
```
Comment (KatieO4t1qbit) [inline ccvm_simulators/ccvmplotlib/tests/test_BoxQP_metadata.py:58]: rm this comment now that the implementation is in :)

File: ccvm_simulators/ccvmplotlib/tests/test_BoxQP_metadata.py [hunk 5]
```diff
@@ -35,52 +49,69 @@ def test_BoxQP_metadata_invalid_problem_type(self):
...
+        boxqp_metadata.ingest_metadata(self.valid_metadata_filepath)
+
+        self.assertGreater(boxqp_metadata._BoxQPMetadata__batch_size, 0)
+        self.assertGreater(len(boxqp_metadata._BoxQPMetadata__problem_size_list), 0)
 
     def test_ingest_result_data_method_invalid(self):
         """Test BoxQP Metadata class "ingest_result_data" method when invalid
         inputs are given.
 
         # TODO: Implementation
```
Comment (KatieO4t1qbit) [inline ccvm_simulators/ccvmplotlib/tests/test_BoxQP_metadata.py:70]: same here! And there are two more instances below that can go.

File: ccvm_simulators/ccvmplotlib/tests/test_BoxQP_metadata.py [hunk 6]
```diff
@@ -35,52 +49,69 @@ def test_BoxQP_metadata_invalid_problem_type(self):
...
         """
-        pass
+        boxqp_metadata = BoxQPMetadata(self.valid_problem_type)
+        boxqp_metadata.ingest_metadata(self.valid_metadata_filepath)
+
+        self.assertGreater(boxqp_metadata._BoxQPMetadata__batch_size, 0)
+        self.assertGreater(len(boxqp_metadata._BoxQPMetadata__problem_size_list), 0)
```
Comment (KatieO4t1qbit) [inline ccvm_simulators/ccvmplotlib/tests/test_BoxQP_metadata.py:64]: Will these values be the same every time, since we are using some set metadata? If so, let's check the exact value.

**Review by KatieO4t1qbit — APPROVED** [review]
Great work, Chan Woo!
I left a few minor suggestions that won't affect the code itself, so I'll approve it but let me know if anything else changes and you'd like me to take another look :)

**Comment by ChanWooY-1QBit** [comment]
@KatieO4t1qbit I addressed all your comments, but if you can take a quick look on the `test_BoxQP_metadata.py` file `example/ccvmp_boxqp_plot.py` files, then it would be great. Thanks!

File: examples/ccvm_boxqp_plot.py [hunk 7]
```diff
@@ -50,20 +52,55 @@
...
+        tts_plot_fig, tts_plot_ax
+    )  # apply default styling
     plt.show()  # show plot in a new window
 
-    # If PLOT_OUTPUT_DIR not exists, create the path
-    if not os.path.isdir(PLOT_OUTPUT_DIR):
-        os.makedirs(PLOT_OUTPUT_DIR)
-        print("Plot folder doesn't exist yet. Creating: ", PLOT_OUTPUT_DIR)
+    # Save to local
+    tts_plot_fig.savefig(PLOT_OUTPUT_DEST)
```
Comment (KatieO4t1qbit) [inline examples/ccvm_boxqp_plot.py:82]: This seems to be the figure that got saved in the TTS plot file, but it looks like ETS to me. Can you check if this happens on your side?

DL-CCVM_TTS_cpu_plot

**Review by KatieO4t1qbit — CHANGES_REQUESTED** [review]

File: ccvm_simulators/tests/unit/ccvmplotlib/test_BoxQP_metadata.py [hunk 8]
```diff
@@ -35,52 +64,78 @@ def test_BoxQP_metadata_invalid_problem_type(self):
...
+        TTS_plot_data = boxqp_metadata.generate_plot_data(
+            metric_func=self.valid_machine_func
+        )
+
+        self.assertIsInstance(TTS_plot_data, pd.DataFrame)
+        self.assertGreater(TTS_plot_data.size, 0)
 
-    def test_generate_arrays_of_TTS_method_invalid(self):
-        """Test BoxQP Metadata class "generate_arrays_of_TTS" method when
+        ETS_plot_data = boxqp_metadata.generate_plot_data(
```
[the remark this answers is not in the document]
Comment (ChanWooY-1QBit) [reply ccvm_simulators/tests/unit/ccvmplotlib/test_BoxQP_metadata.py:113]: It really depends on what metric function you provide. The same `generate_plot_data` method is used for both ETS and TTS plot data calculation, but the output depends on whether it is the machine time calc function or the energy max calc function.

File: examples/ccvm_boxqp_plot.py (hunk 7 above)
Comment (ChanWooY-1QBit) [reply examples/ccvm_boxqp_plot.py:77]: I am getting the same plots.
ets
tts

File: ccvm_simulators/tests/unit/ccvmplotlib/test_BoxQP_metadata.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (KatieO4t1qbit) [reply ccvm_simulators/tests/unit/ccvmplotlib/test_BoxQP_metadata.py:113]: Oh I see. I had a misunderstanding. I didn't realize that the function spits out the final value of the metric that we will be plotting, I thought it spit out a value that was used as _part_ of the metrics calculation.
Side note: with my new understanding, the name `metric_func` makes far more sense.

I think we might still want to improve clarity for the user, i.e. what if they specify an energy function here but then call TTS formatting? Given our current timelines, I'm happy for that to be addressed in a future PR, and we can ignore this comment for now.

File: examples/ccvm_boxqp_plot.py (hunk 7 above)
Comment (KatieO4t1qbit) [reply examples/ccvm_boxqp_plot.py:77]: Hmm, okay let me try again!

Comment (KatieO4t1qbit) [reply examples/ccvm_boxqp_plot.py:77]: Okay I think I found the issue: both plots get saved to the same place: PLOT_OUTPUT_DEST (which we don't modify):
```
ets_plot_fig.savefig(PLOT_OUTPUT_DEST)
```

I think we need a variable here or a second constant.

File: ccvm_simulators/tests/unit/ccvmplotlib/test_BoxQP_metadata.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (KatieO4t1qbit) [reply ccvm_simulators/tests/unit/ccvmplotlib/test_BoxQP_metadata.py:113]: Made a ticket! We can revisit later :)
https://github.com/1QB-Information-Technologies/ccvm/issues/163

File: examples/ccvm_boxqp_plot.py (hunk 7 above)
Comment (ChanWooY-1QBit) [reply examples/ccvm_boxqp_plot.py]: I separated the ETS and TTS plot output destination. And here are the new plots:
ets
tts

**Review by KatieO4t1qbit — APPROVED** [review]
Awesome, looks good Chan Woo! 🔥
